### PR TITLE
usbus: add descriptor prefix support

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds support to the descriptor generators to prepend descriptors to the current element. This is required to add an interface association descriptor before the CDC ACM control interface descriptor. With this descriptor, Windows should correctly detect a RIOT-based CDC ACM peripheral and use the `USBser.sys` driver out of the box.

### Testing procedure

Test together with #12430, the interface association should always be right before the Communications Abstract modem interface descriptor.

### Issues/PRs references

Required for #12430 

Another PR is required to straighten out the terminology `s/header/descriptor/g` to increase alignment with the USB spec.

### CODE REVIEW

File: sys/include/usb/usbus.h [hunk 1]
```diff
@@ -195,16 +208,17 @@ typedef struct {
         /**
          * @brief USBUS generic header generator generated length
          *
-         * Must return the length of the header that will be generated by
-         * @ref get_header
+         * Must return the length total length of the descriptors that will be
```
Comment (dylad) [inline sys/include/usb/usbus.h]:
```suggestion
         * Must return the total length of the descriptor that will be
```

File: sys/include/usb/usbus.h [hunk 2]
```diff
@@ -195,16 +208,17 @@ typedef struct {
...
          *
          * @param usbus The usbus context
          * @param arg   Additional argument for the header generator
          *
          * @return      Length of the generated header
          */
         size_t (*get_header_len)(usbus_t *usbus, void *arg);
-        size_t fixed_len;           /**< length of the header if it is a fixed length */
+        size_t fixed_len;           /**< total length of the generated
+                                         descripors if they are fixed length */
```
Comment (dylad) [inline sys/include/usb/usbus.h]:
```suggestion
                                         descriptors if they are fixed length */
```

File: sys/usb/usbus/usbus_fmt.c [hunk 3]
```diff
@@ -130,17 +130,36 @@ static size_t _hdrs_config_size(usbus_t *usbus)
     return len;
 }
 
-static inline size_t call_get_header(usbus_t *usbus, usbus_hdr_gen_t *hdr)
+static inline size_t _get_pre_header(usbus_t *usbus, usbus_hdr_gen_t *hdr)
 {
-    return hdr->funcs->get_header(usbus, hdr->arg);
+    return hdr->funcs->fmt_pre_descriptor != NULL
```
Comment (dylad) [inline sys/usb/usbus/usbus_fmt.c]:
```suggestion
    return (hdr->funcs->fmt_pre_descriptor != NULL)
```

File: sys/usb/usbus/usbus_fmt.c [hunk 4]
```diff
@@ -130,17 +130,36 @@ static size_t _hdrs_config_size(usbus_t *usbus)
...
+        ? hdr->funcs->fmt_pre_descriptor(usbus, hdr->arg)
+        : 0;
+}
+
+static inline size_t _get_additional_header(usbus_t *usbus, usbus_hdr_gen_t *hdr)
+{
+    return hdr->funcs->get_header != NULL
```
Comment (dylad) [inline sys/usb/usbus/usbus_fmt.c]:
```suggestion
    (return hdr->funcs->get_header != NULL)
```

File: sys/include/usb/usbus.h (hunk 2 above)
Comment (bergzand) [reply sys/include/usb/usbus.h]: Wait, shouldn't have applied your suggestion, the fixed length must be the sum of the prefixed descriptor and the appended descriptor, so "descriptors" is correct here.

File: sys/include/usb/usbus.h (hunk 1 above)
Comment (bergzand) [reply sys/include/usb/usbus.h]: Same as below, multiple descriptors can be generated

Comment (bergzand) [reply sys/include/usb/usbus.h]: Clarified this a bit

File: sys/include/usb/usbus.h (hunk 2 above)
Comment (bergzand) [reply sys/include/usb/usbus.h]: Also clarified this one a bit

Comment (dylad) [reply sys/include/usb/usbus.h]: what is bothering me here is the missing T

File: sys/include/usb/usbus.h (hunk 1 above)
Comment (dylad) [reply sys/include/usb/usbus.h]: Alright

File: sys/include/usb/usbus.h (hunk 2 above)
Comment (bergzand) [reply sys/include/usb/usbus.h]: > what is bothering me here is the missing T

I might have accidentally fixed that with my fixups

**Comment by bergzand** [comment]
Squashed

**Comment by bergzand** [comment]
> Tested on SAME54 with cdc_acm, and lsusb. Everything looks sane.

Thanks for reviewing!

**Pull request merged**